File: cpp/libs/formula/src/dump.cpp
#include "adl2/formula/dump.hpp"

#include "adl2/formula/encode.hpp"

#include <sstream>

namespace adl2::formula {
namespace {

std::string dump_atom(const LinAtom& a) {
  std::ostringstream os;
  os << "(atom [";
  bool first = true;
  for (const auto& t : a.terms()) {
    if (!first) os << " ";
    first = false;
    os << "(" << t.first.dump() << " " << t.second.to_string() << ")";
  }
  os << "] " << rel_str(a.rel()) << " " << a.constant().dump() << ")";
  return os.str();
}

std::string dump_f(const Formula& f) {
  switch (f.kind) {
    case Formula::Kind::True:
      return "true";
    case Formula::Kind::False:
      return "false";
    case Formula::Kind::Atom:
      return dump_atom(f.atom);
    case Formula::Kind::And: {
      std::string s = "(and";
      for (const auto& x : f.items) {
        s += " ";
        s += dump_f(x);
      }
      s += ")";
      return s;
    }
    case Formula::Kind::Or: {
      std::string s = "(or";
      for (const auto& x : f.items) {
        s += " ";
        s += dump_f(x);
      }
      s += ")";
      return s;
    }
    case Formula::Kind::Unknown:
      return "(unknown " + f.diag.to_string() + ")";
    case Formula::Kind::Dual:
      return "(dual " + f.diag.to_string() + " " + dump_f(*f.plus) + " " +
             dump_f(*f.minus) + ")";
  }
  return "?";
}

std::string dump_q(const QFormula& f) {
  switch (f.kind) {
    case QFormula::Kind::True:
      return "true";
    case QFormula::Kind::False:
      return "false";
    case QFormula::Kind::Atom:
      return dump_atom(f.atom);
    case QFormula::Kind::And: {
      std::string s = "(and";
      for (const auto& x : f.items) {
        s += " ";
        s += dump_q(x);
      }
      s += ")";
      return s;
    }
    case QFormula::Kind::Or: {
      std::string s = "(or";
      for (const auto& x : f.items) {
        s += " ";
        s += dump_q(x);
      }
      s += ")";
      return s;
    }
  }
  return "?";
}

}  // namespace

std::string dump_formula(const Formula& f) { return dump_f(f); }
std::string dump_qformula(const QFormula& f) { return dump_q(f); }

std::string dump_encoded(const adl2::sema::Hir& hir,
                         const std::vector<EncodedRegion>& regions) {
  std::ostringstream os;
  os << "unit: " << hir.unit << "\n";
  for (const auto& r : regions) {
    os << "region " << r.region << " " << r.name
       << " exact=" << (r.is_exact() ? "true" : "false") << "\n";
    os << "formula " << dump_f(r.formula) << "\n";
    os << "over " << dump_q(r.formula.over().qformula()) << "\n";
    os << "under " << dump_q(r.formula.under().qformula()) << "\n";
    for (std::size_t i = 0; i < r.diags.size(); ++i) {
      const auto& d = r.diags.entries()[i];
      os << "diag D" << i << " " << d.span.start << ".." << d.span.end << " "
         << d.reason << "\n";
    }
  }
  return os.str();
}

}  // namespace adl2::formula

File: reimplementation/smash3/crates/adl-formula/src/dump.rs
//! Canonical formula dump (P3 oracle form). Byte-for-byte vs smash2_cpp
//! `check --dump-formula`.
//!
//! Rat formatting uses existing `to_parts()` only — this module must not
//! change `adl_sema::Rat` or the encoder.

use crate::encode::EncodedRegion;
use crate::formula::{Formula, QFormula};
use crate::lin::LinAtom;
use adl_sema::{Hir, Rat};

fn dump_rat(r: &Rat) -> String {
    let p = r.to_parts();
    let sign = if p.negative { "-" } else { "" };
    if p.denominator == "1" {
        format!("{sign}{}", p.numerator)
    } else {
        format!("{sign}{}/{}", p.numerator, p.denominator)
    }
}

fn dump_atom(a: &LinAtom) -> String {
    let mut s = String::from("(atom [");
    for (i, (c, q)) in a.terms().iter().enumerate() {
        if i > 0 {
            s.push(' ');
        }
        s.push_str(&format!("({} {q})", dump_rat(c)));
    }
    s.push_str(&format!(
        "] {} {})",
        a.rel().as_str(),
        dump_rat(a.constant())
    ));
    s
}

fn dump_f(f: &Formula) -> String {
    match f {
        Formula::True => "true".into(),
        Formula::False => "false".into(),
        Formula::Atom(a) => dump_atom(a),
        Formula::And(v) => {
            let mut s = String::from("(and");
            for x in v {
                s.push(' ');
                s.push_str(&dump_f(x));
            }
            s.push(')');
            s
        }
        Formula::Or(v) => {
            let mut s = String::from("(or");
            for x in v {
                s.push(' ');
                s.push_str(&dump_f(x));
            }
            s.push(')');
            s
        }
        Formula::Unknown(d) => format!("(unknown {d})"),
        Formula::Dual { plus, minus, why } => {
            format!("(dual {why} {} {})", dump_f(plus), dump_f(minus))
        }
    }
}

fn dump_q(f: &QFormula) -> String {
    match f {
        QFormula::True => "true".into(),
        QFormula::False => "false".into(),
        QFormula::Atom(a) => dump_atom(a),
        QFormula::And(v) => {
            let mut s = String::from("(and");
            for x in v {
                s.push(' ');
                s.push_str(&dump_q(x));
            }
            s.push(')');
            s
        }
        QFormula::Or(v) => {
            let mut s = String::from("(or");
            for x in v {
                s.push(' ');
                s.push_str(&dump_q(x));
            }
            s.push(')');
            s
        }
    }
}

/// Canonical dump of one formula (sexpr).
#[must_use]
pub fn dump_formula(f: &Formula) -> String {
    dump_f(f)
}

/// Canonical dump of a projected QFormula.
#[must_use]
pub fn dump_qformula(f: &QFormula) -> String {
    dump_q(f)
}

/// Canonical per-unit dump: `unit:` header, then per-region formula/over/under/diag.
#[must_use]
pub fn dump_encoded(hir: &Hir, regions: &[EncodedRegion]) -> String {
    let mut s = format!("unit: {}\n", hir.unit);
    for r in regions {
        s.push_str(&format!(
            "region {} {} exact={}\n",
            r.region,
            r.name,
            if r.is_exact() { "true" } else { "false" }
        ));
        s.push_str(&format!("formula {}\n", dump_f(&r.formula)));
        s.push_str(&format!("over {}\n", dump_q(r.formula.over().qformula())));
        s.push_str(&format!("under {}\n", dump_q(r.formula.under().qformula())));
        for (i, d) in r.diags.iter() {
            s.push_str(&format!(
                "diag {i} {}..{} {}\n",
                d.span.start, d.span.end, d.reason
            ));
        }
    }
    s
}
